python/ray/data/_internal/planner/plan_write_op.py

```diff
@@ -9,31 +9,50 @@
     MapTransformer,
 )
 from ray.data._internal.logical.operators.write_operator import Write
-from ray.data.block import Block
+from ray.data.block import Block, BlockAccessor
 from ray.data.datasource.datasink import Datasink
 from ray.data.datasource.datasource import Datasource
 
 
 def generate_write_fn(
     datasink_or_legacy_datasource: Union[Datasink, Datasource], **write_args
 ) -> Callable[[Iterator[Block], TaskContext], Iterator[Block]]:
+    def fn(blocks: Iterator[Block], ctx) -> Iterator[Block]:
+        """Writes the blocks to the given datasink or legacy datasource.
+
+        Outputs the original blocks to be written."""
+        if isinstance(datasink_or_legacy_datasource, Datasink):
+            original_blocks = datasink_or_legacy_datasource.write(blocks, ctx)
+        else:
+            original_blocks = datasink_or_legacy_datasource.write(
+                blocks, ctx, **write_args
+            )
+
+        return original_blocks
```

Comment:
This would be cleaner 
```
it1, it2 = itertools.tee(blocks, n=2)
datasink_or_legacy_datasource.write(it2, ctx)
return it2
```
just let `write` return None

Reply:
One downside of this approach is that blocks cannot be GC'ed until the write is complete. 
This doesn't matter for now, because the blocks are not loaded in a streaming way either.

Reply:
If we're making `Datasink.write` return `None` or the original blocks, we should also change the interface for `Datasink.on_write_complete`. Currently, it assumes that the return values of `Dataset.write` or passed to `on_write_complete`:
https://github.com/ray-project/ray/blob/2b94a682c121ca84774c3661cff6ff325f01ef24/python/ray/data/datasource/datasink.py#L42-L43

Reply:
IMO returning `None` is probably cleanest. 

The write results are only used in `_FileDatasink.on_write_complete` (we'd need to refactor the implementation accordingly if we change the return type).